src/relay/pass/quantize.cc

```diff
@@ -278,24 +570,72 @@ Expr Conv2dRealize(const Call& ref_call,
   if (lhs->dtype != cfg->dtype_input) {
     ldata = Cast(ldata, cfg->dtype_input);
   }
-  Expr rdata = Cast(rhs->data, cfg->dtype_weight);
+  Expr rdata = rhs->data;
+  rdata = Cast(rdata, cfg->dtype_weight);
 
   const auto ref_attrs = ref_call->attrs.as<Conv2DAttrs>();
   auto attrs = make_node<Conv2DAttrs>();
   *attrs = *ref_attrs;
   DataType out_dtype = cfg->dtype_activation;
   attrs->out_dtype = out_dtype;
 
+  Expr input_scale = lhs->dom_scale;
+  Expr weight_scale = rhs->dom_scale;
+
+  Array<IndexExpr> data_shape = input_scale.as<ConstantNode>()->tensor_type()->shape;
+  Array<IndexExpr> weight_shape = weight_scale.as<ConstantNode>()->tensor_type()->shape;
+  CHECK(data_shape.size() == 1);
+  CHECK(weight_shape.size() == 1);
+  size_t data_dim = data_shape[0].as<IntImm>()->value;
+  size_t weight_dim = weight_shape[0].as<IntImm>()->value;
+
+  Expr dom_scale;
+  /* Special handling for strided_slice is needed because it changes the number
+   * of channel dimensions and the number of per-channel scales. We may consider
+   * changing the srided_slice rewrite to something other than identity to avoid
+   * this issue.*/
+  if (data_dim == weight_dim) {
+    // TODO)eqy): special handling for only layer wise scale (when both scales are size 1), we can skip this
+    // calculation and only do the old style:
+    auto* data_scale_tensor = input_scale.as<ConstantNode>();
+    auto* weight_scale_tensor = weight_scale.as<ConstantNode>();
+
+    // CURRENT scheme relies product and weight scales to be matched after
+    // multiplying
+    float max_output_scale =\
+    reinterpret_cast<float*>(data_scale_tensor->data->data)[0]*\
+    reinterpret_cast<float*>(weight_scale_tensor->data->data)[0];
+
+    for (size_t i = 0; i < weight_dim; i++) {
+      float cur_output_scale =\
+      reinterpret_cast<float*>(data_scale_tensor->data->data)[i]*\
+      reinterpret_cast<float*>(weight_scale_tensor->data->data)[i];
+      CHECK(cur_output_scale == max_output_scale);
```

Comment:
We can't compare two float number equal in this way. Also what it meaning of this check?

Reply:
This is really to compare two float values are strictly bitwise equal (or all the bytes here are equal), with the assumption that each scale is a power of two so that multiplication happens perfectly. You are right that this is not generally valid if we allow free non-power of 2 scales.
The meaning of the check is to compare that the scales match in the middle of the accumulation during convolution. `max_output_scale` is just a crude heuristic; we use the max of possible accumulation scales across all the channels during calibration. This check tests whether the `match_scales` pass during calibration worked correctly. There is likely more fine-tuning potential here, but we found that `max` gives the most consistent results.